Implement eBPF native redirect functionality (#12831)
cilium-team/cilium: Documentation/gettingstarted/local-redirect-policy.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _local_redirect_policy:

*****************************
Local Redirect Policy (beta)
*****************************

> This document explains how to configure Cilium's Local Redirect Policy, that
> enables pod traffic, destined to a Layer3 address and Layer port and protocol tuple
> or Kubernetes service, to be redirected locally to a backend pod within a node.
> The CiliumLocalRedirectPolicy is configured as a `CustomResourceDefinition`.
> CiliumLocalRedirectPolicy is namespace-aware, while CiliumClusterwideLocalRedirectPolicy
> is a cluster-scoped version that specifies cluster-wide policies.

Reviewer: Minor wording nits, in particular `Layer port` which doesn't make sense:

```suggestion
This document explains how to configure Cilium's Local Redirect Policy, that
enables pod traffic destined to an IP address and port/protocol tuple
or Kubernetes service to be redirected locally to a backend pod within a node.
The CiliumLocalRedirectPolicy is configured as a `CustomResourceDefinition`.
CiliumLocalRedirectPolicy is namespace-aware, while CiliumClusterwideLocalRedirectPolicy
is a cluster-scoped version that specifies cluster-wide policies.
```

     --namespace kube-system

Verify that Cilium agent pod is running.

.. code-block:: bash

    $ kubectl -n kube-system get pods -l k8s-app=cilium
    NAME           READY   STATUS    RESTARTS   AGE
    cilium-5ngzd   1/1     Running   0          3m19s

Validate that the Cilium Local Redirect Policy CRD has been registered.

.. code-block:: bash

	   $ kubectl get crds
	   NAME                              CREATED AT
	   [...]
	   ciliumlocalredirectpolicies.cilium.io              2020-08-24T05:31:47Z

Create a Cilium Local Redirect Policy Custom Resource
======================================================
There are two types of configurations supported in the ``CustomResource`` in
order to match the traffic that needs to be redirected.

AddressMatcher
---------------

This type of configuration is specified using IP address and Layer 4 port and
protocol.
The example shows how to redirect from traffic matching, IP address ``169.254.169.254``
and Layer 4 port ``8080`` with protocol ``TCP``, to a backend pod deployed with
labels ``app=proxy`` and Layer 4 port ``80`` with protocol ``TCP``.

.. literalinclude:: ../../examples/kubernetes-local-redirect/lrp-addrmatcher.yaml

> .. code-block:: bash

Reviewer: nit: It's a bit confusing but sphinx highlighting for `bash` is different from `shell-session` so broadly these should use `shell-session` for highlighting.

In this specific case however, it's using `|SCM_WEB|` which means this particular case must use `parsed-literal::`, otherwise the URL will not be converted into a real URL at docs build time.

You can use `make render-docs` or `make render-docs-live-preview` to generate and view the docs locally to double-check aspects like this.

    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/lrp-addrmatcher.yaml

Create this custom resource.

.. code-block:: bash

>     $ kubectl apply -f lrp-addrmatcher.yaml

Reviewer: The lines above already create this resource don't they?
```suggestion
```

Verify that this is created.

.. code-block:: bash

    $ kubectl get ciliumlocalredirectpolicies | grep lrp-addr
    NAME           AGE
    lrp-addr       20h

Deploy a backend pod where the traffic needs to be redirected to. The metadata
labels and container port and protocol match with the labels, port and protocol
fields specified in the CiliumLocalRedirectPolicy custom resource created
above, respectively.

.. literalinclude:: ../../examples/kubernetes-local-redirect/backend-pod.yaml

> .. code-block:: bash

Reviewer: ```suggestion
.. parsed-literal::
```

    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/backend-pod.yaml

Verify that the pod is running.

.. code-block:: bash

    $ kubectl get pods | grep lrp-pod
    lrp-pod                      1/1     Running   0          46s

Deploy a client pod that will invoke a curl command to the IP address and port
configuration specified in the ``lrp-addr`` custom resource above.

> .. code-block:: bash

Reviewer: ```suggestion
.. parsed-literal::
```

    $ kubectl create -f \ |SCM_WEB|\/examples/kubernetes-dns/dns-sw-app.yaml
    $ kubectl get po
    NAME                             READY   STATUS    RESTARTS   AGE
    pod/mediabot                     1/1     Running   0          14s

.. code-block:: bash

    $ kubectl exec mediabot -- curl -I -s http://169.254.169.254:8080/index.html
    HTTP/1.1 200 OK
    Server: nginx/1.19.2
    Date: Fri, 28 Aug 2020 01:33:34 GMT
    Content-Type: text/html
    Content-Length: 612
    Last-Modified: Tue, 11 Aug 2020 14:50:35 GMT
    Connection: keep-alive
    ETag: "5f32b03b-264"
    Accept-Ranges: bytes

Verify that the traffic was redirected to the ``lrp-pod`` that was deployed.

.. code-block:: bash

    $ kubectl describe pod lrp-pod  | grep 'IP:'
    IP:           10.16.70.187

.. parsed-literal::

>     sudo tcpdump -i any -n port 80

Reviewer: ```suggestion
    $ sudo tcpdump -i any -n port 80
```

configuration.

.. literalinclude:: ../../examples/kubernetes-local-redirect/lrp-svcmatcher.yaml

.. code-block:: bash

    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/lrp-svcmatcher.yaml

Verify that the custom resource is created.

.. code-block:: bash

    $ kubectl get ciliumlocalredirectpolicies | grep svc
    NAME               AGE
    lrp-svc   20h

Deploy the Kubernetes service for which traffic needs to be redirected.

.. literalinclude:: ../../examples/kubernetes-local-redirect/k8s-svc.yaml

.. code-block:: bash

    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/k8s-svc.yaml

Verify that the service is created.

.. code-block:: bash

    $ kubectl get service | grep 'my-service'
    my-service   ClusterIP   172.20.0.51   <none>        80/TCP     2d7h

Deploy a backend pod where the traffic needs to be redirected to. The metadata
labels and container port and protocol match with the labels, port and protocol
fields specified in the CiliumLocalRedirectPolicy custom resource created
above, respectively.

.. literalinclude:: ../../examples/kubernetes-local-redirect/backend-pod.yaml

> .. code-block:: bash

Reviewer: ```suggestion
.. parsed-literal::
```

    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/backend-pod.yaml

Verify that the pod is running.

.. code-block:: bash

    $ kubectl get pods | grep lrp-pod
    lrp-pod                      1/1     Running   0          46s

Deploy a client pod that will invoke a curl command to the Cluster IP of

> ``my-service`` specified in the ``lrp-svc`` custom resource above.

Reviewer: I wonder if it's better to consolidate the backend / client pod creation towards the start of the doc after Cilium pod creation so it doesn't have to be duplicated for each section?

.. code-block:: bash

>     $ kubectl exec nginx-client -- curl -I -s http://172.20.0.51:80/index.html

Reviewer: There's no `nginx-client` pod, did you mean `mediabot`?

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _local_redirect_policy:

*****************************
Local Redirect Policy (beta)
*****************************

This document explains how to configure Cilium's Local Redirect Policy, that
enables pod traffic, destined to a Layer3 address and Layer port and protocol tuple
or Kubernetes service, to be redirected locally to a backend pod within a node.
The CiliumLocalRedirectPolicy is configured as a `CustomResourceDefinition`.
CiliumLocalRedirectPolicy is namespace-aware, while CiliumClusterwideLocalRedirectPolicy
is a cluster-scoped version that specifies cluster-wide policies.

.. include:: ../beta.rst

.. include:: k8s-install-download-release.rst

Deploy Cilium

> ==================================

Reviewer: nit: The title and the include of the k8s install steps are usually in the reverse order.

The Cilium Local Redirect Policy feature is enabled by default on new deployments
via Helm:

.. parsed-literal::

   helm install cilium |CHART_RELEASE| \\
     --namespace kube-system

Verify that Cilium agent pod is running, and check its status using the

> ``cilium status`` CLI command.

Reviewer: Is there something specific the user needs to look for?

The goal with getting started guides is typically to keep the steps as short and simple as possible. If there's something we need the user to check here, then let's spell that out clearly. If not, I think we can drop this step.

Validate that the Cilium Local Redirect Policy CRD has been registered.

> .. parsed-literal::

Reviewer: nit: FWIW, `parsed-literal` is necessary only for snippets that include pipe-wrapped `|SUBSTITUTIONS|` like `|CHART_RELEASE|`, `|SCM_WEB|`. For all other snippets it's sufficient to either use `::` or `.. code-block:: shell-session`.

(That said, I doubt it will make a huge difference here so I don't mind leaving all of these instances as-is)

Reviewer: Also comparing against https://docs.cilium.io/en/v1.8/gettingstarted/dns/ , the `kubectl` lines begin with `$` to indicate that this is a shell command, and to help differentiate the command from the output to make that a bit clearer.

	   kubectl get crds
	   NAME                              CREATED AT
	   [...]
	   ciliumlocalredirectpolicies.cilium.io              2020-08-24T05:31:47Z

Create a Cilium Local Redirect Policy Custom Resource
======================================================
There are two types of configurations supported in the ``CustomResource`` in
order to match the traffic that needs to be redirected.

AddressMatcher
This type of configuration is specified using IP address and Layer 4 port and
protocol.
The example shows how to redirect from traffic matching, IP address ``169.254.169.254``
and Layer 4 port ``8080`` with protocol ``TCP``, to a backend pod deployed with
labels ``app=proxy`` and Layer 4 port ``80`` with protocol ``TCP``.

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumLocalRedirectPolicy
    metadata:
      name: "lrp-metadata"
    spec:
      redirectFrom:
        addressMatcher:
          ip: "169.254.169.254"
          portProtocol:
            port: "8080"
            protocol: TCP
      redirectTo:
        localEndpointSelector:
          matchLabels:
            app: proxy
        portProtocol:
          port: "80"

>           protocol: TCP

Reviewer: If they're not already somewhere in the tree, I think it would make sense to include these files directly in `examples/` somewhere, maybe `examples/local-redirect/lrp-proxy.yaml`? Then you can use `.. literalinclude ../../examples/foo` to insert the YAML into the doc.

Other GSGs will typically do this, then ask the user to create it using `kubectl create -f \ |SCM_WEB|\/examples/path/to/foo.yaml`. That way it's all pulling from a single source, the actual YAML in the tree.

With this in place, these same files can then also be consumed by the CI for validating the functionality, so we know that the CI validates the examples in the tree.

Create this custom resource.

.. parsed-literal::

    kubectl apply -f clrp.yaml

Verify that this is created.

.. parsed-literal::

    kubectl get ciliumlocalredirectpolicies
    NAME               AGE
    lrp-metadata       20h

Now deploy a backend pod where the traffic needs to be redirected to. The metadata
labels and container port and protocol match with the labels, port and protocol
fields specified in the CiliumLocalRedirectPolicy custom resource created above.

.. code-block:: yaml

    apiVersion: v1
    kind: Pod
    metadata:
      name: lrp-pod
      labels:
        app: proxy
    spec:
      containers:
      - name: lrp-pod
        image: nginx
        ports:
        - containerPort: 80
          protocol: TCP

Verify that the pod is running.

.. parsed-literal::

    kubectl get pods | grep lrp-pod
    lrp-pod                      1/1     Running   0          46s

Deploy a client pod that will invoke a curl command to the IP address and port
configuration specified in the CiliumLocalRedirectPolicy CR above in order to match

> traffic that needs to be redirected.

Reviewer: Where does the client come from? Please provide a YAML to deploy so that the set of steps can be easily copy/pasted to follow the guide.

Reviewer: I snuck over to the DNS policy GSG and applied mediabot from there and was able to confirm the functionality locally via that:
```
$ kubectl create -f https://raw.githubusercontent.com/cilium/cilium/v1.8/examples/kubernetes-dns/dns-sw-app.yaml
```

    01:36:24.608600 IP 10.16.70.187.80 > 10.16.215.55.60876: Flags [S.], seq 1315636594, ack 2119454274, win 27960, options [mss 1410,sackOK,TS val 2962246962 ecr 2541637677,nop,wscale 7], length 0
    01:36:24.608638 IP 10.16.215.55.60876 > 10.16.70.187.80: Flags [.], ack 1, win 221, options [nop,nop,TS val 2541637677 ecr 2962246962], length 0
    01:36:24.608867 IP 10.16.215.55.60876 > 10.16.70.187.80: Flags [P.], seq 1:96, ack 1, win 221, options [nop,nop,TS val 2541637677 ecr 2962246962], length 95: HTTP: HEAD /index.html HTTP/1.1
    01:36:24.608876 IP 10.16.70.187.80 > 10.16.215.55.60876: Flags [.], ack 96, win 219, options [nop,nop,TS val 2962246962 ecr 2541637677], length 0
    01:36:24.609007 IP 10.16.70.187.80 > 10.16.215.55.60876: Flags [P.], seq 1:239, ack 96, win 219, options [nop,nop,TS val 2962246962 ecr 2541637677], length 238: HTTP: HTTP/1.1 200 OK
    01:36:24.609052 IP 10.16.215.55.60876 > 10.16.70.187.80: Flags [.], ack 239, win 229, options [nop,nop,TS val 2541637677 ecr 2962246962], length 0

> 

Reviewer: This section is good for testing and confirming that the traffic in fact went to the right place, but in practice there is nothing else in a test env that listens on the 169.254 address so the fact that you were able to query a page and get a successful result from that IP indicates that the LRP is working. So do we need to perform these steps at all?

Author: I see a value in having this verification step since it confirms that the frontend traffic was indeed redirected to the backend pod selected by the LRP. I would like to get inputs from others as well.

> ServiceMatcher

Reviewer 2: Pardon my ignorance, but I wonder what is an exact real world use case for an LRP with a ServiceMatcher?

Author: > @aditighag Thanks, that helps with the review. Do you know any real world use case for ServiceMatcher? From the short glimpse to https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/, it seems that node-local DNS is accessed by non-ClusterIP.

Node-local dns proxy listens on the clusterIP of kube-dns service.

> ServiceMatcher

Reviewer: Is this `ServiceMatcher` line supposed to be a title?

FWIW we use multiple levels for titles:

```
Main title
==========

Section
-------

Subsection
~~~~~~~~~~
```

This type of configuration is specified using Kubernetes service name and namespace
for which traffic needs to be redirected.
The example shows how to redirect from traffic matching ``my-service``, to a
backend pod deployed with labels ``app=proxy`` and Layer 4 port ``80``
with protocol ``TCP``.

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumLocalRedirectPolicy
    metadata:
      name: "lrp-my-svc"
    spec:
      redirectFrom:
        serviceMatcher:
          service:
            k8sService:
              serviceName: my-service
              namespace: default
          portProtocol:

>             port: tcp

Reviewer: This port doesn't seem right. However when I inserted this LRP, nothing seemed to complain, it just didn't work. Is the validation working correctly?

Is `PortProtocol` really necessary for a service selector? The service itself already provides ports.

Author: Synced up offline.

            protocol: TCP
      redirectTo:
        localEndpointSelector:
          matchLabels:
            app: proxy
        portProtocol:
          port: "80"
          protocol: TCP

Create this custom resource.

.. parsed-literal::

    kubectl apply -f clrp-service.yaml

Verify that this is created.

.. parsed-literal::

    kubectl get ciliumlocalredirectpolicies
    NAME               AGE

>     lrp-my-svc   20h

Reviewer: At this point in the guide, there are two CLRPs because the instructions have not told to remove the first one. Is that OK?

Deploy the Kubernetes service for which the traffic needs to be redirected from.

.. code-block:: yaml

    apiVersion: v1
    kind: Service
    metadata:
      name: my-service
    spec:
      selector:
        app: MyApp
      ports:
        - name: tcp
          protocol: TCP
          port: 80

>           targetPort: 9376

Reviewer: I'm not sure I understand the relationship between this service declaration and the nginx pod below. The selector is different.

I was not able to successfully query this service following the instructions below, likely because of this issue.

Author: We select a backend pod based on the `localEndpointSelector` labels specified in the LRP. I've tested serviceMatcher LRP multiple times, I likely missed some steps here or misconfigured the LRP spec here. Will look into this. Thanks!

Verify that the pod is running.

.. parsed-literal::

    kubectl get pods | grep lrp-pod
    lrp-pod                      1/1     Running   0          46s

Check that the service is created.

.. parsed-literal::

    kubectl get service | grep 'my-service'
    my-service   ClusterIP   172.20.0.51   <none>        80/TCP     2d7h

Deploy a client pod that will invoke a curl command to the Cluster IP of
``my-service`` specified in the CiliumLocalRedirectPolicy CR above in order to match
traffic that needs to be redirected.

.. parsed-literal::

    kubectl exec nginx-client -- curl -I -s http://172.20.0.51:80/index.html
    HTTP/1.1 200 OK
    Server: nginx/1.19.2
    Date: Fri, 28 Aug 2020 01:50:50 GMT
    Content-Type: text/html
    Content-Length: 612
    Last-Modified: Tue, 11 Aug 2020 14:50:35 GMT
    Connection: keep-alive
    ETag: "5f32b03b-264"
    Accept-Ranges: bytes

Limitations
==================================
When you create a CiliumLocalRedirectPolicy, traffic for all the new connections
that get established after the policy is enforced will be redirected. But if
you have existing active connections to remote pods that match the configurations

> specified in the policy, then these might not get redirected.

Reviewer: nit: Might make sense to add a short summary line of how to ensure that all such connections are redirected? For instance:

> To ensure all such connections are redirected locally, restart the client pods after configuring the `CiliumLocalRedirectPolicy`.
